main/src/cgeo/geocaching/utils/FilterUtils.java

```diff
@@ -0,0 +1,97 @@
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public class FilterUtils {
+
+    private FilterUtils() {
+        // should not be instantiated
+    }
+
+    public static void openFilterActivity(final Activity activity, final GeocacheFilterContext filterContext, final Collection<Geocache> filteredList) {
+
+        GeocacheFilterActivity.selectFilter(
+            activity,
+            filterContext,
+            filteredList, true);
+    }
+
+    public static <T extends Activity & FilteredActivity> boolean openFilterList(final T filteredActivity, final GeocacheFilterContext filterContext) {
+        final List<GeocacheFilter> filters = new ArrayList<>(GeocacheFilter.Storage.getStoredFilters());
+
+        if (filters.isEmpty()) {
+            return false;
+        } else {
+            final boolean isFilterActive = filterContext.get().isFiltering();
+            if (isFilterActive) {
+                SimpleDialog.of(filteredActivity).setTitle(R.string.cache_filter_storage_select_clear_title)
+                    .setButtons(0, 0, R.string.cache_filter_storage_clear_button)
+                    .setSelectionForNeutral(false)
+                    .selectSingle(filters, (f, pos) -> TextParam.text(f.getName()), -1, false,
+                        (f, pos) -> filteredActivity.refreshWithFilter(f),
+                        (f, pos) -> {
+                        },
+                        (f, pos) -> filteredActivity.refreshWithFilter(GeocacheFilter.createEmpty())
+                    );
+            } else {
+                SimpleDialog.of(filteredActivity).setTitle(R.string.cache_filter_storage_select_title)
+                    .selectSingle(filters, (f, pos) -> TextParam.text(f.getName()), -1, false,
+                        (f, pos) -> filteredActivity.refreshWithFilter(f),
+                        (f, pos) -> {
+                        }
+                    );
+            }
+        }
+        return true;
+    }
+
+    public static void updateFilterBar(final Activity activity, final Collection<String> filterNames) {
+        final View filterView = activity.findViewById(R.id.filter_bar);
+        if (filterNames.isEmpty()) {
+            filterView.setVisibility(View.GONE);
+        } else {
+            final TextView filterTextView = activity.findViewById(R.id.filter_text);
+            filterTextView.setText(TextUtils.join(", ", filterNames));
+            filterView.setVisibility(View.VISIBLE);
+        }
+    }
+
+    /** filterView must exist */
+    public static void initializeFilterBar(@NonNull final Activity activity, @NonNull final FilteredActivity filteredActivity) {
+        final View filterView = activity.findViewById(R.id.filter_bar);
+        filterView.setOnLongClickListener(v -> filteredActivity.showFilterList(null));
+    }
+
+    public static void initializeFilterMenu(@NonNull final Activity activity, @NonNull final FilteredActivity filteredActivity) {
```

Comment:
I did not get the CGeoMap to work with `<T extends Activity & FilteredActivity>`
so, for the other activities it looks strange, calling it with `initializeFilterMenu(this, this)`

@eddiemuc probably you have a solution? 
Passing the view for menu_filter does not work (because of the Runnable handler), it will be null then on passing it to initializeFilterMenu

Reply:
Right, in CgeoMap the activity class and FilteredActivity class are probably different classes hm? Then the approach will not work of course. So better don't use it.

Apart from that I am not sure what the problem is?

Reply:
No problem then. Was just not really sure about that.